# Validation Basics
>  This chapter focuses on the basics of model validation. From splitting data into training, validation, and testing datasets, to creating an understanding of the bias-variance tradeoff, we build the foundation for the techniques of K-Fold and Leave-One-Out validation practiced in chapter three.

- toc: true 
- badges: true
- comments: true
- author: Lucas Nunes
- categories: [Python, Datacamp, Machine Learning]
- image: images/datacamp/1_supervised_learning_with_scikit_learn/2_regression.png

> Note: This is a summary of the course's chapter 2 exercises "Model Validation in Python" at datacamp. <br>[Github repo](https://github.com/lnunesAI/Datacamp/) / [Course link](https://www.datacamp.com/tracks/machine-learning-scientist-with-python)

In [15]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_absolute_error as mae

plt.rcParams['figure.figsize'] = (8, 8)

## Creating train, test, and validation datasets

### Create one holdout set

<div class=""><p>Your boss has asked you to create a simple random forest model on the <code>tic_tac_toe</code> dataset. She doesn't want you to spend much time selecting parameters; rather she wants to know how well the model will perform on future data. For future Tic-Tac-Toe games, it would be nice to know if your model can predict which player will win. </p>
<p>The dataset <code>tic_tac_toe</code> has been loaded for your use.</p>
<p>Note that in Python, <code>=\</code> indicates the code was too long for one line and has been split across two lines.</p></div>

In [ ]:
tic_tac_toe = pd.read_csv('https://raw.githubusercontent.com/lnunesAI/Datacamp/main/2-machine-learning-scientist-with-python/11-model-validation-in-python/datasets/tic-tac-toe.csv')

Instructions
<ul>
<li>Create the <code>X</code> dataset by creating dummy variables for all of the categorical columns.</li>
<li>Split <code>X</code> and <code>y</code> into train (<code>X_train</code>, <code>y_train</code>) and test (<code>X_test</code>, <code>y_test</code>) datasets.</li>
<li>Split the datasets using 10% for testing</li>
</ul>

In [ ]:
# Create dummy variables using pandas
X = pd.get_dummies(tic_tac_toe.iloc[:,0:9])
y = tic_tac_toe.iloc[:, 9]

# Create training and testing datasets. Use 10% for the test set
X_train, X_test, y_train, y_test  = train_test_split(X, y, test_size=0.1, random_state=1111)

**Remember, without the holdout set, you cannot truly validate a model. Let's move on to creating two holdout sets.**

### Create two holdout sets

<div class=""><p>You recently created a simple random forest model to predict Tic-Tac-Toe game wins for your boss, and at her request, you did not do any parameter tuning. Unfortunately, the overall model accuracy was too low for her standards. This time around, she has asked you to focus on model performance.</p>
<p>Before you start testing different models and parameter sets, you will need to split the data into training, validation, and testing datasets. Remember that after splitting the data into training and testing datasets, the validation dataset is created by splitting the training dataset.</p>
<p>The datasets <code>X</code> and <code>y</code> have been loaded for your use.</p></div>

Instructions
<ul>
<li>Create temporary datasets and testing datasets (<code>X_test</code>, <code>y_test</code>). Use 20% of the overall data for the testing datasets. </li>
<li>Using the temporary datasets (<code>X_temp</code>, <code>y_temp</code>), create training (<code>X_train</code>, <code>y_train</code>) and validation (<code>X_val</code>, <code>y_val</code>) datasets.</li>
<li>Use 25% of the <strong>temporary data</strong> for the validation datasets.</li>
</ul>

In [ ]:
# Create temporary training and final testing datasets
X_temp, X_test, y_temp, y_test  =\
    train_test_split(X, y, test_size=0.2, random_state=1111)

# Create the final training and validation datasets
X_train, X_val, y_train, y_val =\
    train_test_split(X_temp, y_temp, test_size=0.25, random_state=1111)

**You now have training, validation, and testing datasets, but do you know when you need both validation and testing datasets? Keep going! The next exercise will help make sure you understand when to use validation datasets.**

### Why use holdout sets

<div class=""><p>It is important to understand when you would use three datasets (training, validation, and testing) instead of two (training and testing). There is no point in creating an additional dataset split if you are not going to use it. </p>
<p>When should you consider using training, validation, <em>and</em> testing datasets?</p></div>

When there is a lot of data. Splitting into three sets helps speed up modeling.

**When testing parameters, tuning hyper-parameters, or anytime you are frequently evaluating model performance.**

Only when you are running regression and not classification models.

Only when you are running classification and not regression models


**Anytime we are evaluating model performance repeatedly we need to create training, validation, and testing datasets.**

## Accuracy metrics: regression models

### Mean absolute error

<div class=""><p>Communicating modeling results can be difficult. However, most clients understand that on average, a predictive model was off by some number. This makes explaining the mean absolute error easy. For example, when predicting the number of wins for a basketball team, if you predict 42, and they end up with 40, you can easily explain that the error was two wins.</p>
<p>In this exercise, you are interviewing for a new position and are provided with two arrays. <code>y_test</code>, the true number of wins for all 30 NBA teams in 2017 and <code>predictions</code>, which contains a prediction for each team. To test your understanding, you are asked to both manually calculate the MAE and use <code>sklearn</code>.</p></div>

In [ ]:
#@title ⠀ { display-mode: "form" }
y_test = np.array([53, 51, 51, 49, 43, 42, 42, 41, 41, 37, 36, 31, 29, 28, 20, 67, 61,
       55, 51, 51, 47, 43, 41, 40, 34, 33, 32, 31, 26, 24])
predictions = np.array([60, 62, 42, 42, 30, 50, 52, 42, 44, 35, 30, 30, 35, 40, 15, 72, 58,
       60, 40, 42, 45, 46, 40, 35, 25, 40, 20, 34, 25, 24])

Instructions
<ul>
<li>Manually calculate the MAE using <code>n</code> as the number of observations predicted.</li>
<li>Calculate the MAE using <code>sklearn</code>.</li>
<li>Print off both accuracy values using the print statements.</li>
</ul>

In [ ]:
from sklearn.metrics import mean_absolute_error

# Manually calculate the MAE
n = len(predictions)
mae_one = sum(abs(y_test - predictions)) / n
print('With a manual calculation, the error is {}'.format(mae_one))

# Use scikit-learn to calculate the MAE
mae_two = mean_absolute_error(y_test, predictions)
print('Using scikit-lean, the error is {}'.format(mae_two))

With a manual calculation, the error is 5.9
Using scikit-lean, the error is 5.9


**These predictions were about six wins off on average. This isn't too bad considering NBA teams play 82 games a year. Let's see how these errors would look if you used the mean squared error instead.**

### Mean squared error

In [ ]:
from sklearn.metrics import mean_squared_error

n = len(predictions)
# Finish the manual calculation of the MSE
mse_one = sum((y_test - predictions) ** 2) / n
print('With a manual calculation, the error is {}'.format(mse_one))

# Use the scikit-learn function to calculate MSE
mse_two = mean_squared_error(y_test, predictions)
print('Using scikit-lean, the error is {}'.format(mse_two))

With a manual calculation, the error is 49.1
Using scikit-lean, the error is 49.1


**If you run any additional models, you will try to beat an MSE of 49.1, which is the average squared error of using your model. Although the MSE is not as interpretable as the MAE, it will help us select a model that has fewer 'large' errors.**

### Performance on data subsets

<div class=""><p>In professional basketball, there are two conferences, the East and the West. Coaches and fans often only care about how teams in their own conference will do this year. </p>
<p>You have been working on an NBA prediction model and would like to determine if the predictions were better for the East or West conference. You added a third array to your data called <code>labels</code>, which contains an "E" for the East teams, and a "W" for the West.  <code>y_test</code> and <code>predictions</code> have again been loaded for your use.</p></div>

In [ ]:
labels= np.array(['E', 'E', 'E', 'E', 'E', 'E', 'E', 'E', 'E', 'E', 'E', 'E', 'E',
       'E', 'E', 'W', 'W', 'W', 'W', 'W', 'W', 'W', 'W', 'W', 'W', 'W',
       'W', 'W', 'W', 'W'])

Instructions 1/4
<li>Create an array <code>east_teams</code> that can be used to filter <code>labels</code> to East conference teams.</li>

In [ ]:
# Find the East conference teams
east_teams = labels == 'E'
# West teams
west_teams = labels == 'W'

Instructions 2/4
<li>Create the arrays <code>true_east</code> and <code>preds_east</code> by filtering the arrays <code>y_test</code> and <code>predictions</code>.</li>

In [ ]:
# Create arrays for the true and predicted values
true_east = y_test[east_teams]
preds_east = predictions[east_teams]
# West teams
true_west = y_test[west_teams]
preds_west = predictions[west_teams]

Instructions 3/4
<li>Use the print statements to print the MAE (using <code>scikit-learn</code>) for the East conference. The <code>mean_absolute_error</code> function has been loaded as <code>mae</code>.</li>

In [ ]:
# Print the accuracy metrics
print('The MAE for East teams is {}'.format(
    mae(true_east, preds_east)))

The MAE for East teams is 6.733333333333333


Instructions 4/4
<li>The variable <code>west_error</code> contains the MAE for the West teams. Use the print statement to print out the Western conference MAE.</li>

In [ ]:
# Print the West accuracy
west_error = mae(true_west, preds_west)
print('The MAE for West conference is {}'.format(west_error))

The MAE for West conference is 5.066666666666666


**It looks like the Western conference predictions were about two games better on average. Over the past few seasons, the Western teams have generally won the same number of games as the experts have predicted. Teams in the East are just not as predictable as those in the West.**

## Classification metrics

### Confusion matrices

<div class=""><p>Confusion matrices are a great way to start exploring your model's accuracy. They provide the values needed to calculate a wide range of metrics, including sensitivity, specificity, and the F1-score.</p>
<p>You have built a classification model to predict if a person has a broken arm based on an X-ray image. On the testing set, you have the following confusion matrix:</p>
<table>
<thead>
<tr>
<th></th>
<th>Prediction: 0</th>
<th>Prediction: 1</th>
</tr>
</thead>
<tbody>
<tr>
<td>Actual: 0</td>
<td>324 (TN)</td>
<td>15 (FP)</td>
</tr>
<tr>
<td>Actual: 1</td>
<td>123 (FN)</td>
<td>491 (TP)</td>
</tr>
</tbody>
</table></div>

Instructions
<ul>
<li>Use the confusion matrix to calculate overall accuracy.</li>
<li>Use the confusion matrix to calculate precision and recall. </li>
<li>Use the three print statements to print each accuracy value.</li>
</ul>

In [ ]:
# Calculate and print the accuracy
accuracy = (324  + 491) / (953)
print("The overall accuracy is {0: 0.2f}".format(accuracy))

# Calculate and print the precision
precision = (491) / (15 + 491)
print("The precision is {0: 0.2f}".format(precision))

# Calculate and print the recall
recall = (491) / (123  + 491 )
print("The recall is {0: 0.2f}".format(recall))

The overall accuracy is  0.86
The precision is  0.97
The recall is  0.80


**In this case, a true positive is a picture of an actual broken arm that was also predicted to be broken. Doctors are okay with a few additional false positives (predicted broken, not actually broken), as long as you don't miss anyone who needs immediate medical attention.**

### Confusion matrices, again

<div class=""><p>Creating a confusion matrix in Python is simple. The biggest challenge will be making sure you understand the orientation of the matrix. This exercise makes sure you understand the <code>sklearn</code> implementation of confusion matrices. Here, you have created a random forest model using the <code>tic_tac_toe</code> dataset <code>rfc</code> to predict outcomes of 0 (loss) or 1 (a win) for Player One.</p>
<p><em>Note:</em> If you read about confusion matrices on another website or for another programming language, the values might be reversed.</p></div>

In [3]:
from sklearn.ensemble import RandomForestClassifier
df = pd.read_csv('https://raw.githubusercontent.com/lnunesAI/Datacamp/main/2-machine-learning-scientist-with-python/11-model-validation-in-python/datasets/tic_tac_toe_preprocessed.csv')
X, y = df.iloc[:, :-1], df.iloc[:, -1]
X_train, y_train = X.iloc[:95].values, y.iloc[:95].values
X_test, y_test = X.iloc[95:].values, y.iloc[95:].values

In [8]:
X_train.shape, X_test.shape, y_train.shape

((95, 27), (863, 27), (95,))

In [4]:
rfc = RandomForestClassifier(n_estimators=500, random_state=1111)
rfc.fit(X_train, y_train)

RandomForestClassifier(bootstrap=True, ccp_alpha=0.0, class_weight=None,
                       criterion='gini', max_depth=None, max_features='auto',
                       max_leaf_nodes=None, max_samples=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=500,
                       n_jobs=None, oob_score=False, random_state=1111,
                       verbose=0, warm_start=False)

Instructions
<ul>
<li>Import <code>sklearn</code>'s function for creating confusion matrices.</li>
<li>Using the model <code>rfc</code>, create category predictions on the test set <code>X_test</code>.</li>
<li>Create a confusion matrix using <code>sklearn</code>.</li>
<li>Print the value from <code>cm</code> that represents the actual 1s that were predicted as 1s (true positives).</li>
</ul>

In [5]:
from sklearn.metrics import confusion_matrix

# Create predictions
test_predictions = rfc.predict(X_test)

# Create and print the confusion matrix
cm = confusion_matrix(y_test, test_predictions)
print(cm)

# Print the true positives (actual 1s that were predicted 1s)
print("The number of true positives is: {}".format(cm[1, 1]))

[[143 154]
 [ 39 527]]
The number of true positives is: 527


**Row 1, column 1 represents the number of actual 1s that were predicted 1s (the true positives). Always make sure you understand the orientation of the confusion matrix before you start using it!**

### Precision vs. recall

<div class=""><p>The accuracy metrics you use to evaluate your model should <em>always</em> be based on the specific application. For this example, let's assume you are a really sore loser when it comes to playing Tic-Tac-Toe, but only when you are certain that you are going to win.</p>
<p>Choose the most appropriate accuracy metric, either precision or recall, to complete this example. But remember, <em>if you think you are going to win, you better win!</em></p>
<p>Use <code>rfc</code>, which is a random forest classification model built on the <code>tic_tac_toe</code> dataset.</p></div>

Instructions
<ul>
<li>Import the precision or the recall metric for <code>sklearn</code>. Only <em>one</em> method is correct for the given context. </li>
<li>Calculate the precision or recall using <code>y_test</code> for the true values and <code>test_predictions</code> for the predictions.</li>
<li>Print the final score based on your selected metric.</li>
</ul>

In [6]:
from sklearn.metrics import precision_score

test_predictions = rfc.predict(X_test)

# Create precision or recall score based on the metric you imported
score = precision_score(y_test, test_predictions)

# Print the final result
print("The precision value is {0:.2f}".format(score))

The precision value is 0.77


**Precision is the correct metric here. Sore-losers can't stand losing when they are certain they will win! For that reason, our model needs to be as precise as possible. With a precision of only 79%, you may need to try some other modeling techniques to improve this score.**

## The bias-variance tradeoff

### Error due to under/over-fitting

<div class=""><p>The candy dataset is prime for overfitting. With only 85 observations, if you use 20% for the testing dataset, you are losing a lot of vital data that could be used for modeling. Imagine the scenario where most of the chocolate candies ended up in the training data and very few in the holdout sample. Our model might <em>only</em> see that chocolate is a vital factor, but fail to find that other attributes are also important. In this exercise, you'll explore how using too many features (columns) in a random forest model can lead to overfitting.</p>
<p>A <em>feature</em> represents which columns of the data are used in a decision tree. The parameter <code>max_features</code> limits the number of features available.</p></div>

In [13]:
df = pd.read_csv('https://raw.githubusercontent.com/lnunesAI/Datacamp/main/2-machine-learning-scientist-with-python/11-model-validation-in-python/datasets/candy-data.csv')
X = df.drop(['competitorname', 'winpercent'], axis=1)
y = df['winpercent']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=1111)


from sklearn.ensemble import RandomForestRegressor
rfr = RandomForestRegressor(n_estimators=25, random_state=1111, max_features=2)

Instructions 1/3
<p>Create a random forest model with 25 trees, a random state of 1111, and <code>max_features</code> of 2. Read the print statements.</p>

In [17]:
# Update the rfr model
rfr = RandomForestRegressor(n_estimators=25,
                            random_state=1111,
                            max_features=2)
rfr.fit(X_train, y_train)

# Print the training and testing accuracies 
print('The training error is {0:.2f}'.format(
  mae(y_train, rfr.predict(X_train))))
print('The testing error is {0:.2f}'.format(
  mae(y_test, rfr.predict(X_test))))

The training error is 3.88
The testing error is 9.15


Instructions 2/3
<p>Set <code>max_features</code> to 11 (the number of columns in the dataset). Read the print statements.</p>

In [18]:
# Update the rfr model
rfr = RandomForestRegressor(n_estimators=25,
                            random_state=1111,
                            max_features=11)
rfr.fit(X_train, y_train)

# Print the training and testing accuracies 
print('The training error is {0:.2f}'.format(
  mae(y_train, rfr.predict(X_train))))
print('The testing error is {0:.2f}'.format(
  mae(y_test, rfr.predict(X_test))))

The training error is 3.57
The testing error is 10.05


Instructions 3/3
<p>Set <code>max_features</code> equal to 4. Read the print statements.</p>

In [19]:
# Update the rfr model
rfr = RandomForestRegressor(n_estimators=25,
                            random_state=1111,
                            max_features=4)
rfr.fit(X_train, y_train)

# Print the training and testing accuracies 
print('The training error is {0:.2f}'.format(
  mae(y_train, rfr.predict(X_train))))
print('The testing error is {0:.2f}'.format(
  mae(y_test, rfr.predict(X_test))))

The training error is 3.60
The testing error is 8.79


**The chart below shows the performance at various max feature values. Sometimes, setting parameter values can make a huge difference in model performance.**

![](https://assets.datacamp.com/production/repositories/3981/datasets/7e30218261b88cc6e57da1e07b73c5803450ccf6/Screen%20Shot%202019-01-13%20at%205.40.29%20PM.png)

### Am I underfitting?

<div class=""><p>You are creating a random forest model to predict if you will win a future game of Tic-Tac-Toe. Using the <code>tic_tac_toe</code> dataset, you have created training and testing datasets, <code>X_train</code>, <code>X_test</code>, <code>y_train</code>, and <code>y_test</code>. </p>
<p>You have decided to create a bunch of random forest models with varying amounts of trees (1, 2, 3, 4, 5, 10, 20, and 50). The more trees you use, the longer your random forest model will take to run. However, if you don't use enough trees, you risk underfitting. You have created a for loop to test your model at the different number of trees.</p></div>

In [31]:
df = pd.read_csv('https://raw.githubusercontent.com/lnunesAI/Datacamp/main/2-machine-learning-scientist-with-python/11-model-validation-in-python/datasets/tic-tac-toe.csv')
X = pd.get_dummies(df.iloc[:, 0:9])
y = df.iloc[:, 9]
y = df['Class'].map({'positive': 1, 'negative': 0})
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=1111)

Instructions
<ul>
<li>For each loop, predict values for both the <code>X_train</code> and <code>X_test</code> datasets.</li>
<li>For each loop, append the <code>accuracy_score()</code> of the <code>y_train</code> dataset and the corresponding predictions to <code>train_scores</code>.</li>
<li>For each loop, append the <code>accuracy_score()</code> of the <code>y_test</code> dataset and the corresponding predictions to <code>test_scores</code>.</li>
<li>Print the training and testing scores using the print statements.</li>
</ul>

In [33]:
from sklearn.metrics import accuracy_score

test_scores, train_scores = [], []
for i in [1, 2, 3, 4, 5, 10, 20, 50]:
    rfc = RandomForestClassifier(n_estimators=i, random_state=1111)
    rfc.fit(X_train, y_train)
    # Create predictions for the X_train and X_test datasets.
    train_predictions = rfc.predict(X_train)
    test_predictions = rfc.predict(X_test)
    # Append the accuracy score for the test and train predictions.
    train_scores.append(round(accuracy_score(y_train, train_predictions), 2))
    test_scores.append(round(accuracy_score(y_test, test_predictions), 2))
# Print the train and test scores.
print("The training scores were: {}".format(train_scores))
print("The testing scores were: {}".format(test_scores))

The training scores were: [0.94, 0.93, 0.98, 0.97, 0.99, 1.0, 1.0, 1.0]
The testing scores were: [0.83, 0.79, 0.89, 0.91, 0.91, 0.93, 0.97, 0.98]


**Notice that with only one tree, both the train and test scores are low. As you add more trees, both errors improve. Even at 50 trees, this still might not be enough. Every time you use more trees, you achieve higher accuracy. At some point though, more trees increase training time, but do not decrease testing error.**